### PR TITLE
Continuing to pull out matrix function implementations from MatrixBase

### PR DESCRIPTION
**Pull request by tom-pytel**
#### References to other Issues or PRs
Similar to #18342, #18420, #18425 but now working on MatrixBase

#### Brief description of what is fixed or changed
Simplifying `matrices.py` by pulling function implementations out into separate source files and cleaning up a little,

#### Release Notes

* matrices
  * Pulled out and cleaned up implementeations of upper/lower_triangular_solve, cholesky decomposition and cholesky_solve.

### CODE REVIEW

**Comment by Pristine-Cat** [comment]
Well I started pulling relevant functions out of `MatrixBase` into two new source files `matrices/decompositions.py` and `matrices/solvers.py`. This is just a start for this PR but I wanted to send it up to see if you guys agree with the way I am doing it.

`MatrixBase` contains a LOT of function implementations and it is unlike the previous classes I have pulled functions out of in that many of these have specific dense and sparse implementations (in many cases with incompatible function signatures). What I am doing is basically relocating the implementations of both the dense and sparse versions into the respective source file which describes their functionality (so far `decompositions` and `solvers` but will also add `inverse` eventually).

I am collapsing the number of functions to clean up the matrix namespace a little, for example there is no longer a public `lower_triangular_solve` and private `_lower_triangular_solve` helper. This is because there are so many of these split functions that the number of matrix methods becomes excessive (I think). The downside to that is that when calling `lower_triangular_solve` from internal functions some basic checks will be duplicated but I am balancing this with lowering the method count. Though in the case of `cholesky_solve` this worries me a little since it may wind up checking `is_hermitian` twice, once in the solve and once in `cholesky` so would like an opinion on this.

Unlike the previous classes I am fixing small inconsistencies in the functions along the way because there are so many of them that I would probably lose track of them by the time everything is pulled out and I do a cleanup pass. The small fixes so far are:

* `as_mutable().copy()` changed to `as_mutable()` since this already returns a copy.
* TypeError -> ShapeError in `upper_triangular_solve` for consistency with `lower_triangular_solve`.
* The implementation of `_cholesky_sparse` did not have the `hermitian` argument like dense `cholesky`, added this and the appropriate check.

The methods which are different between dense and sparse matrices are added in `DenseMatrix` and `SparseMatrix` like the helper functions were previously and disappear from `MatrixBase` (also opinion on this pls).

**Comment by Pristine-Cat** [comment]
Hmm, configuration error? I can't restart the test myself:
```
sympy/printing/tests/test_theanocode.py[38] ................/home/travis/miniconda/envs/test-environment/lib/python3.6/site-packages/theano/gof/opt.py:251: UserWarning: WARNING: Supervisor is not added. Please build a FunctionGraphvia theano.compile.function_module.std_graph()or add the Supervisor class manually.
```

**Comment by oscarbenjamin** [comment]
> 
> sympy/printing/tests/test_theanocode.py[38] ................/home/travis/miniconda/envs/test-environment/lib/python3.6/site-packages/theano/gof/opt.py:251: UserWarning: WARNING: Supervisor is not added. Please build a FunctionGraphvia theano.compile.function_module.std_graph()or add the Supervisor class manually.
> 

That warning is always there. I'm not sure what it means.

BTW one you can restart the tests is by closing and reopening the PR.

File: sympy/matrices/decompositions.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+                L[i, j] = dps((1 / L[j, j])*(M[i, j] -
+                    sum(L[i, k]*L[j, k] for k in range(j))))
+
+            L[i, i] = sqrt(dps(M[i, i] -
+                sum(L[i, k]**2 for k in range(i))))
+
+    return M._new(L)
+
+
+def _cholesky_sparse(M, hermitian=True, dotprodsimp=None):
```
Comment (sylee957) [inline sympy/matrices/decompositions.py]: I think that `hermitian=True` should be the decomposition `L * L.H` unlike documented here. 
And I don't think that it's a correct patch unless you add an algorithm to compute hermitian cholesky for sparse.

File: sympy/matrices/decompositions.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    [ 3, 3, 0],
+    [-1, 1, 3]])
+    >>> A.cholesky() * A.cholesky().T == A
+    True
+    """
+
+    from sympy.core.numbers import nan, oo
+
+    def cholesky_sparse(M, dotprodsimp=None):
+        """Algorithm for numeric Cholesky factorization of a sparse matrix."""
```
Comment (sylee957) [inline sympy/matrices/decompositions.py]: It looks too long to put inside the function.

File: sympy/matrices/decompositions.py (hunk 1 above)
Comment (Pristine-Cat) [reply sympy/matrices/decompositions.py]: Ok, for some stupid reason I had the impression the sparse algorithm could handle it but obviously not. Added the hermitian changes to the algorithm and now it behaves like dense:
```julia
>>> Matrix([[5 + I, 0], [0, 1]]).cholesky(hermitian=False)
Matrix([
[sqrt(5 + I), 0],
[          0, 1]])
>>> _ * _.T
Matrix([
[5 + I, 0],
[    0, 1]])

>>> SparseMatrix([[5 + I, 0], [0, 1]]).cholesky(hermitian=False)
Matrix([
[sqrt(5 + I), 0],
[          0, 1]])
>>> _ * _.T
Matrix([
[5 + I, 0],
[    0, 1]])

>>> Matrix([[2, 1 + I], [1 - I, 2]]).cholesky(hermitian=True)
Matrix([
[                sqrt(2), 0],
[sqrt(2)/2 - sqrt(2)*I/2, 1]])
>>> simplify (_ * _.H)
Matrix([
[    2, 1 + I],
[1 - I,     2]])

>>> SparseMatrix([[2, 1 + I], [1 - I, 2]]).cholesky(hermitian=True)
Matrix([
[                sqrt(2), 0],
[sqrt(2)/2 - sqrt(2)*I/2, 1]])
>>> simplify (_ * _.H)
Matrix([
[    2, 1 + I],
[1 - I,     2]])
```

**Comment by Pristine-Cat** [comment]
Pulled out LDL stuff and other minor tweaks and fixes. Also deleted this from `MatrixBase`:
```julia
def __getattr__(self, attr):
    if attr in ('diff', 'integrate', 'limit'):
        def doit(*args):
            item_doit = lambda item: getattr(item, attr)(*args)
            return self.applyfunc(item_doit)

        return doit
    else:
        raise AttributeError(
            "%s has no attribute %s." % (self.__class__.__name__, attr))
```
That looks like legacy stuff from before `MatrixCalculus` since those methods are defined there and this will never be reached.

Added `hermitian` support to `_LDLdecomposition_sparse` since it was not there before just like `_cholesky_sparse`. The tests seem to tell me I did it right and my brain also thinks so but you might want to double check that I added it correctly.

**Comment by Pristine-Cat** [comment]
Continuing to pull out decompositions and solvers, I think only `QRdecomposition` left in this PR before I would like to merge it gets too big, then on to inverses which are even more messy than anything before.

**Comment by Pristine-Cat** [comment]
Changes here:
* Put reference back into LUdecompositionFF docstring.
* Pulled out QRdecomposition and QRsolve.
* Pulled out gauss_jordan_solve.
* Added dotprodsimp to LUdecomposition (to pass on to LUdecomposition_Simple) and QR stuff.

I think this is a good place to merge before the next PR. I had left the `dotprodsimp` stuff out of QRdecomposition and QRsolve before since the docstring said it was mainly for "educational purposes" but it does make a big difference in some cases so I added it:
```julia
$ python -m timeit -r1 -n1 -s "from sympy import Matrix, Symbol; x = Symbol('
x'); M = Matrix ([[x+1, x-1, x+1], [x-1, x+1, x-1], [x+1, x-1, x+1]])" "M.QRdecomposition(dotprodsimp=False)"
1 loops, best of 1: 7.98 sec per loop
```
vs.
```julia
$ python -m timeit -r1 -n1 -s "from sympy import Matrix, Symbol; x = Symbol('
x'); M = Matrix ([[x+1, x-1, x+1], [x-1, x+1, x-1], [x+1, x-1, x+1]])" "M.QRdecomposition(dotprodsimp=True)"
1 loops, best of 1: 1.35 sec per loop
```

**Comment by sylee957** [comment]
Generally, I'd not try to put up much meanings for 'educational' for computer algebra, because everything that are not purely numeric would not be practical. So I'd agree on proceeding with adding the simplify.

**Comment by Pristine-Cat** [comment]
So if there are no change requests this is good to merge.

**Comment by oscarbenjamin** [comment]
This looks fine to me. @sylee957 what do you think?

**Pull request merged**